02/app.js

```diff
@@ -1,32 +1,39 @@
-document.addEventListener('DOMContentLoaded', init);
+document.addEventListener("DOMContentLoaded", init);
 
 function init() {
-    const divList = document.querySelectorAll('div');
-    
-    setBorderColorAsync(divList[0], 'red', function() {
-        setBorderColorAsync(divList[1], 'blue', function() {
-            setBorderColorAsync(divList[2], 'green', function() {
-                console.log('finish');
-            });
-        });
-    });
+    const divList = document.querySelectorAll("div");
+    const colors = ["red", "blue", "green"];
 
+    setBorderColorSequentially(Array.from(divList), colors)
+        .then(() => {
+            console.log("finish");
+        })
+        .catch((error) => {
+            console.error(error);
+        });
 }
 
-function setBorderColorAsync(element, color, callback) {
-    if(element && element instanceof HTMLElement) {
-        // sprawdzam czy parametr jest elementem DOM, więcej:
-        // https://stackoverflow.com/questions/384286/javascript-isdom-how-do-you-check-if-a-javascript-object-is-a-dom-object
-        
-        if(callback && typeof callback === 'function') {
+function setBorderColorAsync(element, color) {
```

Comment:
Jeśli mamy już funkcje z obietnicami to chodziło o proste rozwiązanie tj. pokazanie, że obietnicę można wywoływać jedną po drugiej tj.
```
setBorderColorAsync(divList[0], 'red')
    .then(() => setBorderColorAsync(divList[1], 'blue'))
    .then(() => setBorderColorAsync(divList[2], 'green'))
    .then(() => console.log('finish'))
```

Twoje rozwiązanie jest jak najbardziej ok, ale zdecydowanie za bardzo "wymasterowane" :P